ocrd_tesserocr/recognize.py

```diff
@@ -281,3 +120,157 @@ def process(self):
+                                           underlined=None if 'underlined' not in word_attributes else word_attributes['underlined'],
+                                           monospace=None if 'monospace' not in word_attributes else word_attributes['monospace'],
+                                           serif=None if 'serif' not in word_attributes else word_attributes['serif']
+                                           )
+                word.set_TextStyle(word_style) # (or somewhere in custom attribute?)
+            # word_type = result_it.BlockType()
+            # PT.UNKNOWN
+            # PT.FLOWING_TEXT
+            # PT.HEADING_TEXT
+            # PT.PULLOUT_TEXT
+            # PT.EQUATION
+            # PT.TABLE
+            # PT.VERTICAL_TEXT
+            # PT.CAPTION_TEXT
+            # PT.HORZ_LINE
+            # PT.VERT_LINE
+            # PT.NOISE
+            # PT.COUNT
+            # ...
+            # add word annotation unconditionally (i.e. even for glyph level):
+            word.add_TextEquiv(TextEquivType(Unicode=result_it.GetUTF8Text(RIL.WORD), conf=result_it.Confidence(RIL.WORD)/100))
+            if maxlevel == 'word':
+                # maybe add TextEquiv alternatives via ChoiceIterator for WORD?
+                pass
+            else:
+                self._process_glyphs_in_word(word, result_it)
+            if result_it.IsAtFinalElement(RIL.TEXTLINE, RIL.WORD):
+                break
+            else:
+                result_it.Next(RIL.WORD)
+
+    def _process_existing_words(self, words, maxlevel, tessapi):
+        for word in words:
+            log.debug("Recognizing text in word '%s'", word.id)
+            word_xywh = xywh_from_points(word.get_Coords().points)
+            tessapi.SetRectangle(word_xywh['x'], word_xywh['y'], word_xywh['w'], word_xywh['h'])
+            tessapi.SetPageSegMode(PSM.SINGLE_WORD)
+            if word.get_TextEquiv():
+                log.warning("Word '%s' already contains text results", word.id)
+            word_conf = tessapi.AllWordConfidences()
+            word_conf = word_conf[0]/100.0 if word_conf else 0.0
+            word.add_TextEquiv(TextEquivType(Unicode=tessapi.GetUTF8Text().rstrip("\n\f"), conf=word_conf))
+            if maxlevel == 'word':
+                # maybe add TextEquiv alternatives via ChoiceIterator for WORD?
+                continue # next word (to avoid indentation below)
+            ## glyph level:
+            glyphs = word.get_Glyph()
+            if glyphs:
+                ## external glyph layout:
+                # raise Exception("existing annotation for Glyph level would clash with OCR results for word '%s'", word.id) # forcing external layout annotation for gylphs is worse with Tesseract
+                log.warning("Word '%s' contains glyphs already, recognition might be suboptimal", word.id)
+                for glyph in glyphs:
```

Comment:
I really like the idea of how you split up `self.process` into levels and "existing" (external) vs non-existing (internal) structure. Maybe for consistency and clarity, at this point one could likewise separate #L233 through #L248 into a `self._process_existing_glyphs(glyphs, tessapi)`?

Reply:
Good idea, done.